sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala

```diff
@@ -473,6 +476,27 @@ class FileBasedDataSourceSuite extends QueryTest with SharedSQLContext with Befo
       }
     }
   }
+
+  test("SPARK-25237 compute correct input metrics in FileScanRDD") {
+    withTempPath { p =>
+      val path = p.getAbsolutePath
+      spark.range(1000).repartition(1).write.csv(path)
+      val bytesReads = new mutable.ArrayBuffer[Long]()
+      val bytesReadListener = new SparkListener() {
+        override def onTaskEnd(taskEnd: SparkListenerTaskEnd) {
+          bytesReads += taskEnd.taskMetrics.inputMetrics.bytesRead
+        }
+      }
+      sparkContext.addSparkListener(bytesReadListener)
+      try {
+        spark.read.csv(path).limit(1).collect()
+        sparkContext.listenerBus.waitUntilEmpty(1000L)
+        assert(bytesReads.sum === 7860)
```

Comment:
So the sum *should* be 10*2 + 90*3 + 900*4 = 3890. That's the size of the CSV file that's written too, when I try it locally. When I run this code without the change here, I get 7820+7820 = 15640. So this is better! but I wonder why it ends up thinking it reads about twice the bytes?

Reply:
In this test, Spark run with `local[2]` and each scan thread points to the same CSV file. Since each thread gets the file size thru Hadoop APIs, the total `byteRead` becomes 2 * the file size, IIUC.

Reply:
7860/2=3930, 40 bytes more than expected, but I'm willing to believe there's a good reason for that somewhere in how it gets read. Clearly it's much better than the answer of 15640, so willing to believe this is fixing something.

Reply:
yea, actually the file size is `3890`, but the hadoop API (`FileSystem.getAllStatistics
) reports that number (`3930`). I didn't look into the Hadoop code yet, so I don't get why. I'll dig into it later.

Reply:
> In this test, Spark run with local[2] and each scan thread points to the same CSV file. Since each thread gets the file size thru Hadoop APIs, the total byteRead becomes 2 * the file size, IIUC.

I am afraid it's not that case, csv will infer schema first, which will try to load the the first row in the path, then the actually read. That's why the input bytes read is doubled. It may be more reasonable to just write and read `text` file.

As for `3930 = 3890 + 40`, the extra 40 bytes is the crc file size. Hadoop uses `ChecksumFileSystem` internally.

And one more thing: this test case may be inaccurate. If the task completes successfully, all the data is consumed, `updateBytesReadWithFileSize` is a no-op, and `updateBytesRead()` in the close function will update the correct size.
FYI @maropu

Reply:
Ah, I see. Can you make a pr to fix that?